rust/template/differential_datalog/program/mod.rs

```diff
@@ -1595,39 +1601,41 @@ impl Program {
     }
 
     /// Returns all input relations of the program
-    fn input_relations(&self) -> Vec<RelId> {
-        self.nodes
-            .iter()
-            .flat_map(|node| match node {
-                ProgNode::Rel { rel: r } => {
-                    if r.input {
-                        vec![r.id]
-                    } else {
-                        vec![]
-                    }
+    fn input_relations<'a>(&'a self) -> impl Iterator<Item = RelId> + 'a {
+        self.nodes.iter().filter_map(|node| match node {
+            ProgNode::Rel { rel: r } => {
+                if r.input {
+                    Some(r.id)
+                } else {
+                    None
                 }
-                ProgNode::Apply { .. } => vec![],
-                ProgNode::SCC { rels: rs } => {
-                    for r in rs {
-                        assert!(!r.rel.input, "input relation ({}) in SCC", r.rel.name);
-                    }
-                    vec![]
+            }
+            ProgNode::Apply { .. } => None,
+            ProgNode::SCC { rels: rs } => {
+                for r in rs {
+                    assert!(!r.rel.input, "input relation ({}) in SCC", r.rel.name);
                 }
-            })
-            .collect()
+
+                None
+            }
+        })
     }
 
     /// Return all relations required to compute rels, excluding recursive dependencies on rels
-    fn dependencies(rels: &[&Relation]) -> FnvHashSet<Dep> {
+    fn dependencies<'a, R>(mut rels: R) -> FnvHashSet<Dep>
```

Comment:
This where this PR breaks things.  The original version takes a slice and traverses it multiple times.  The new version takes iterator.  It exhausts the iterator in the first loop, so subsequent call to `rels.all` always returns `true`, so `filter` does nothing.

Reply:
@Kixiron , let me know if you'll have time to fix this; otherwise I can do it.

Reply:
> This is not performance-critical code, but your improvements still make things cleaner.

I take it back. It turns out that `struct Program` can grow huge for large programs due all the strings that are further bloated by type annotations injected by the type checker. I really like this PR now as it avoids cloning this struct. I am going to fix the bug and merge.

Reply:
Actually, it's the other PR that does this, and it still copies the program ones.

Reply:
I believe that `self: Arc<Self>` is valid in stable rust, that would allow not cloning the physical program any

Reply:
Good point. I'll try that too.  I am also going to change types of all strings in `Program` to `Cow<String>`, as these are typically static strings, so we don't need to allocate them even once.